utils/like_matcher.cc

```diff
@@ -0,0 +1,97 @@
+
+/*
+ * Copyright 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include "like_matcher.hh"
+
+#include <boost/locale/encoding.hpp>
+#include <string>
+
+namespace {
+
+using std::wstring;
+
+/// Processes a new pattern character, extending re with the equivalent regex pattern.
+void process_char(wchar_t c, wstring& re, bool& escaping) {
```

Comment:
I don't understand why `escaping` is passed by reference. You always overwrite with `false` at the end of the function.

Reply:
@debesb lines 36-37. It's not hard to drop the reference by returning a value from process_char,
~~~
bool should_escape = false;
while (...) {
    should_escape = process_char(c, re, should_escape);
}
~~~
though I'm not 100% sure if it's more or less readable.

Reply:
My point is that the value of `escaping` never changes between calls. It is always assigned `false` at the end of `process_char()`, as well as before the first call (in the caller).

Reply:
But it does! line 36 sets it to true and line 37 returns. It's used to change the state before the next `process_char()` call in a loop (line 83).

Reply:
Oh, I missed that.

I personally prefer `should_escape = process_char(c, re, should_escape)`, I'll leave it up to Dejan to change or leave as-is.

Reply:
Leaving as-is.  This is exactly like `mbstate_t`, which is passed by reference to various library functions.